### DISCUSSION TITLE
How to build with different configurations?

### DISCUSSION THREAD

**Discussion by informarte** [com-lihaoyi/mill#1441] [Q&A] [answered]
I am looking for a way to produce development and release builds, the latter using inlining. Are there examples (in terms of mill-based open-source projects) that could serve as a blueprint?

**Comment by lefou** [comment, marked as answer]
I assume, you want to apply special/different `scalacOptions` in the release builds, but otherwise want to the same project sources and settings. I can't point you to existing example open source projects right now, but this kind of setup is easy to realize with mill. Of course there are multiple ways to accomplish this. My preferred solution is to have two almost identical sub modules.

```scala
object myapp extends module {
  trait Settings extends ScalaModule {
    override def millSourcePath = myapp.millSourcePath
    // common settings and dependencies
  }
  object dev extends Settings {
    // development-specific settings
  } 
  object prod extends Settings {
    // production-specific settings
    override def scalacOptions = // ..., e.g. enable inlining
  }
}
```

If you want to make one of the two configurations the default you could also let `myapp` directly extends the shared settings:

```scala
trait Settings extends ScalaModule {
    // common settings and dependencies
}
object myapp extends Settings {
  // development-specific settings

  object prod extends Settings {
    override def millSourcePath = myapp.millSourcePath
    // production-specific settings
    override def scalacOptions = // ..., e.g. enable inlining
  }
}
```

What are the alternatives?

* You could create a cross module to differentiate the build targets. This is probably a bit harder to understand for non-regular mill users.
* You could switch the settings depending on some environment variable. Although, this is the way other tools like `make` or Maven would prefer, it is not very mill idiomatic.

**Reply by informarte** [reply]
A follow-up question regarding IntelliJ IDEA: With the above setup, `./mill mill.scalalib.GenIdea/idea` generates two modules, namely `myapp.dev` and `myapp.prod`. When loading the project, the IDE chooses one of them as the current module for building. Is there a way to set the current module, either in the IDE or with some setting in the `build.sc` file?

**Reply by lefou** [reply]
You can skip the generation for IDEA for specific modules with `def skipIdea = true`.

**Reply by informarte** [reply]
Thanks for the workaround!
